scalalib/src/mill/scalalib/JavaModule.scala

```diff
@@ -455,13 +504,62 @@ trait JavaModule
   /**
    * Returns a function adding BOM and dependency management details of
    * this module to a `coursier.core.Dependency`
+   *
+   * Meant to be called on values originating from `ivyDeps`
    */
   def processDependency(
       overrideVersions: Boolean = false
   ): Task[coursier.core.Dependency => coursier.core.Dependency] = Task.Anon {
-    val bomDeps0 = allBomDeps().toSeq.map(_.withConfig(Configuration.compile))
+    val bomDeps0 =
+      allBomIvyDeps().toSeq.map(_.withConfig(Configuration.compile)) ++ extraBomIvyDeps().toSeq
+    val depMgmt = processedDependencyManagement(
+      (extraDepManagement().toSeq ++ depManagement().toSeq)
+        .map(bindDependency())
+        .map(_.dep)
+    )
+
+    addBoms(bomDeps0, depMgmt, overrideVersions = overrideVersions)
```

Comment:
Can we refactor `processDependency`, `processCompileDependency`, and `processRunDependency` to share the logic between them via an anonymous task. That will help make it easier to see the core differences between them.

I'm also not clear about the differences here:

- `processCompileDependency` uses `compileDepManagement()`, and `processRunDependency` uses `runDepManagement`
- `processDependency` and `processCompileDependency` both use `extraDepManagement` but `processRunDependency` does not. Why?
- Only `processDependency` seems to use `extraBomIvyDeps`, while `processCompileDependency` and `processRunDependency` do not. Why?

Reply:
This should also simplify the way BOMs are handled when taking scopes into account. I've been having issues reconciling Mill's `compileIvyDeps` / `ivyDeps` / `runIvyDeps` (where scopes are known beforehand) with how scopes are handled in BOMs (not known beforehand, dependencies from all scopes are mixed, and BOMs can change the scope if it's undefined).